Don't treat numbers larger than the JS max number size as number values in environment variables

skizer opened the pull request:
-Fixes incorrect parsing of the ```env. file``` with numeric values that are outside of ```Number.MAX_SAFE_INTEGER``` resulting in unwanted behaviour. (e.g -> oauth -> discord_client_id)
-Removed unnecessary multiple "IF" statements since value can only be ether one of the listed values.

P.S This is my first ever pull request on github, please bare with me when I've missed something out.

Commit:
Incorrect parsing of numeric values in env.

* Fixes incorrect parsing of the env. file with numeric values that are outside of Number.MAX_SAFE_INTEGER resulting in unwanted behaviour.
- Like wrong client_ids for oauth. (tested with discord oauth)
* Removed unnecessary multiple "IF" statements since value can only be ether one of the listed values.

Commit:
Merge pull request #1 from skizer/env-oauth-parsing-fix

Incorrect parsing of numeric values in env.

paescuj commented:
@skizer Good catch, thanks! Would it be okay for you if I'm going to include this change in my pull request #6101 as I'm working on the exact same lines? Props to you of course 😃

skizer commented:
@paescuj Ya sure, I'm fine with that :)
I've also recently noticed that you can't use custom_params inside the .env file which grant supports to pass additionally parameters to the oauth2 provider.

e.g discord allows to set the "prompt" parameter to skip the next authorization without the users need of approving it first.

Commit:
Implements custom_param for OAUTH via grant

According to grants documentation you can provide additionally custom parameters to supported OAUTH provider with ```custom_params```.
This change allows to add them in JSON format and thus adding multiple parameters.

paescuj commented:
@skizer Great, thanks! I've included the check for `Number.MAX_SAFE_INTEGER` in pull request #6101.

I'm not sure about your other change regarding JSON values (@rijkvanzanten):
- It doesn't seem like custom params for OAUTH is implemented at the moment (at least not mentioned in the docs), right?
- Do we even want JSON support in env variables?
- Do we need a better check, e.g. try to convert to JSON beforehand? (Error handling)

Depending on @rijkvanzanten opinion here, it might be the best to open up a new pull request or issue for this.

skizer commented:
> @skizer Great, thanks! I've included the check for `Number.MAX_SAFE_INTEGER` in pull request #6101.

@paescuj You're welcome!
> I'm not sure about your other change regarding JSON values (@rijkvanzanten):
>
> * It doesn't seem like custom params for OAUTH is implemented at the moment (at least not mentioned in the docs), right?

This is fine, since directus soly relies on grant for the OAUTH flow, thus making this implementation working right away.
That's also why the docs reference grant if you're interested into the available options. The doc needs an update on that anyway if this gets merged. Also probably a rework due many people getting confused over what they can do in directus with OAUTH. (People asking in discord how oauth in directus works, what they can do and why it doesn't work)
> * Do we even want JSON support in env variables?

Exactly, IMO usually such stuff shouldn't even be configured in the env file due the nature of different types as parameters. Then again there's the pro for me that all your config stuff lives in one location, which speaks for it again :'D 

> * Do we need a better check, e.g. try to convert to JSON beforehand? (Error handling)

No, since the console will scream at you right when trying to start up directus - misconfigured stuff should always fail, otherwise it leads to unwanted behaviour and in worst case even not noticing when going for production.

> Depending on @rijkvanzanten opinion here, it might be the best to open up a new pull request or issue for this.

Can do that, let's wait for @rijkvanzanten :)

paescuj commented:
Thanks for your feedback!

> This is fine, since directus soly relies on grant for the OAUTH flow, thus making this implementation working right away.
That's also why the docs reference grant if you're interested into the available options. The doc needs an update on that anyway if this gets merged but probably only because of many people getting confused over what they can do in directus with OAUTH. (People asking in discord how oauth in directus works, what they can do and why it doesn't work)
> 
> If you're going straight according to the docs this would even be the correct way, it would say you should look up the grant doc and see what options are available and see that the custom_params actually expects this datatype.

So all OAUTH environment variables are (already) getting passed to `grant`?

> Exactly, IMO usually such stuff shouldn't even be configured in the env file due the nature of different types as parameters. Then again there's the pro for me that all your config stuff lives in one location, which speaks for it again :'D

Yeah, there are pros and downs... That's why I've raised this question 😄

> No, since the console will scream at you right when trying to start up directus - misconfigured stuff should always fail, otherwise it leads to unwanted behaviour and in worst case even not noticing when going for production.

I agree!

rijkvanzanten commented:
> P.S This is my first ever pull request on github, please bare with me when I've missed something out.

Could've fooled me! You're a natural at this

> I've also recently noticed that you can't use custom_params inside the .env file which grant supports to pass additionally parameters to the oauth2 provider.

Any parameter you set is passed on to Grants configuration object lowercased:

https://github.com/directus/directus/blob/792281ceb656519679dfa0e8808724de4e77711e/api/src/grant.ts#L27-L35

That being said, this should be updated to work the same as DB_ and various other prefixes to allow for nested fields (https://docs.directus.io/reference/environment-variables/#type-casting-and-nesting) using `getConfigFromEnv`:

https://github.com/directus/directus/blob/792281ceb656519679dfa0e8808724de4e77711e/api/src/database/index.ts#L17

> Do we even want JSON support in env variables?

I don't think we need it at this point! The `getConfigFromEnv` setup we use in other places already supports nested values, so the custom params for Grant should be configurable with the same setup.

> No, since the console will scream at you right when trying to start up directus - misconfigured stuff should always fail, otherwise it leads to unwanted behaviour and in worst case even not noticing when going for production.

100% agree

rijkvanzanten commented:
Lets consider the nested value support for Grant config a separate PR, makes it a little cleaner in the archiving / release logs etc 🙂

rijkvanzanten committed:
Fix linter warnings, remove JSON support in favor of nested Grant support
